flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -484,77 +484,158 @@ public boolean isShutdown() {
-						ev.getTaskNameWithSubtaskIndex(),
-						job);
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+		try {
+			// make some eager pre-checks
+			synchronized (lock) {
+				preCheckBeforeTriggeringCheckpoint(isPeriodic, props.forceCheckpoint());
 			}
-		}
 
-		// we will actually trigger this checkpoint!
+			final Execution[] executions = getTriggerExecutions();
+			final Map<ExecutionAttemptID, ExecutionVertex> ackTasks = getAckTasks();
+
+			// we will actually trigger this checkpoint!
+			Preconditions.checkState(!isTriggering);
+			isTriggering = true;
+
+			// keep a reference outside the async calls
+			final AtomicReference<PendingCheckpoint> currentCheckpointRef = new AtomicReference<>();
+			initializeCheckpoint(props, externalSavepointLocation)
+				.thenComposeAsync((checkpointIdAndStorageLocation) -> {
+					synchronized (lock) {
+						// since we haven't created the PendingCheckpoint yet, we need to check the
+						// global state here.
+						// otherwise the PendingCheckpoint would be discarded if the global state
+						// changed (shutdown or periodic scheduling stopped)
+						if (shutdown) {
+							throw new CompletionException(new CheckpointException(CheckpointFailureReason.CHECKPOINT_COORDINATOR_SHUTDOWN));
+						}
+						if (isPeriodic && !periodicScheduling) {
+							throw new CompletionException(new CheckpointException(CheckpointFailureReason.PERIODIC_SCHEDULER_SHUTDOWN));
+						}
+					}
+					final PendingCheckpoint pendingCheckpoint = createPendingCheckpoint(
+						timestamp,
+						props,
+						ackTasks,
+						isPeriodic,
+						checkpointIdAndStorageLocation.checkpointId,
+						checkpointIdAndStorageLocation.checkpointStorageLocation,
+						onCompletionPromise);
+					currentCheckpointRef.set(pendingCheckpoint);
+
+					LOG.info("Triggering checkpoint {} @ {} for job {}.",
+						checkpointIdAndStorageLocation.checkpointId, timestamp, job);
+					return snapshotMasterState(pendingCheckpoint);
+				}, timer)
+				.thenRun(() -> {
+					// it must be executed in timer thread.
+					// we could add a checking after replace timer with main thread
+					final PendingCheckpoint checkpoint = currentCheckpointRef.get();
+					checkState(!checkpoint.isDiscarded());
+					snapshotTaskState(
+						timestamp,
+						checkpoint.getCheckpointId(),
+						checkpoint.getCheckpointStorageLocation(),
+						props,
+						executions,
+						advanceToEndOfTime);
+					onTriggerSuccess();
+				})
+				.whenCompleteAsync((ignored, throwable) -> {
+					// the error handling has to be executed in timer thread
+					if (throwable != null) {
+						// the initialization might not be finished yet
+						if (currentCheckpointRef.get() == null) {
+							onTriggerFailure(onCompletionPromise, throwable);
+						} else {
+							onTriggerFailure(currentCheckpointRef.get(), throwable);
+						}
+					}
+				}, timer);
```

Comment:
> Why are you splitting this into a three chain calls after initializeCheckpoint? Can not this be initializeCheckpoint -> "everything else at once" ?

Because snapshotting master hooks is also asynchronous. So there is a `thenComposeAsync` + `thenRun`. The last `whenCompleteAsync` is for error handling of all these async chains. Actually `exceptionally` will be easier to read. However the error handling must be executed in timer thread, there is no such an `exceptionallyAsync` available.

> Also final AtomicReference<PendingCheckpoint> currentCheckpointRef = new AtomicReference<>(); seems strange having to use AtomicReference for something that's basically a synchronous chain of calls one after another, right?

The `AtomicReference` is used to refer the `PendingCheckpoint` in lambda expressions (chain calls and error handling). It has to be a final variable. Let me try to make it more readable.

Reply:
So it's the following chain of actions:

1. `initializeCheckpoint` in the `ioExecutor`
2. `createPendingCheckpoint` in the `timer`
3. `triggerHook` in the `ioExecutor`
4. follow up of `triggerHook` (lines 694:717 from CheckpointCoordinator) in `timer`
5.`snapshotTaskState` and `onTriggerFailure` in the `timer`?

If we squashed 4 and 5, wouldn't that simplify the code?

Reply:
The part 4 is quite independent. And I think someday the part 4 could be put to another place, like being executed concurrently with snapshotting task state, or even being removed.
So I think we should make it as an independent method.

I reorganized this piece of codes a bit to make it shorter. I also removed the `AtomicReference` outside the chain calls.

Reply:
BTW, I replaced the `checkState(!checkpoint.isDiscarded());` by `!checkpoint.isDiscarded()`. Because I found there is a possibility that `PendingCheckpoint` might be cancelled during the asynchronous call if there is no master hook (otherwise `snapshotMasterState` would check state of `PendingCheckpoint`).